pyopencl/tools.py

```diff
@@ -463,6 +463,13 @@ def add_dtype(self, dtype):
         if dtype in pyopencl.cltypes.vec_type_to_scalar_and_count:
             return
 
+        if hasattr(dtype, "subdtype") and dtype.subdtype is not None:
+            if dtype.subdtype[0] in [np.float64 or np.complex128]:
```

Comment:
This should probably recurse instead of inlining a partial check.

Reply:
(Alternatively, check that this is a primitive `dtype` and raise `NotImplementedError` otherwise.)

Reply:
done it recursively